pandas/_libs/tslib.pyx

```diff
@@ -607,11 +607,11 @@ cpdef array_to_datetime(
                             # to check if all arguments have the same tzinfo
                             tz = py_dt.utcoffset()
 
-                        except (ValueError, OverflowError):
+                        except (ValueError, OverflowError) as err:
                             if is_coerce:
                                 iresult[i] = NPY_NAT
                                 continue
-                            raise TypeError("invalid string coercion to datetime")
+                            raise type(err)(f"invalid string coercion to datetime for \"{val}\" at position {i}")
```

Comment:
hi @mroeschke , i had a query regarding this -is raising this as always TypeError excepted ? 
in current dev branch code, we are having multiple similar exception being caught & re raised from this flow.
example --> This `TypeError` is raised here --> it is caught below as part of `except TypeError` & `_array_to_datetime_object(values, errors, dayfirst, yearfirst)` is invoked --> this invokation results in the same exception of incorrect data being raised again.

Reply:
example of such an exception flow
```
>>> pd.to_datetime(pd.Series(['2016-01-01', '12']), errors='raise')
Traceback (most recent call last):
  File "pandas/_libs/tslib.pyx", line 603, in pandas._libs.tslib.array_to_datetime
    py_dt = parse_datetime_string(val,
  File "pandas/_libs/tslibs/parsing.pyx", line 264, in pandas._libs.tslibs.parsing.parse_datetime_string
    raise ValueError(f'Given date string {date_string} not likely a datetime.')
ValueError: Given date string 12 not likely a datetime.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pandas/_libs/tslib.pyx", line 614, in pandas._libs.tslib.array_to_datetime
    raise TypeError(f"invalid string coercion to datetime for \"{val}\" at position {i}")
TypeError: invalid string coercion to datetime for "12" at position 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pandas/_libs/tslib.pyx", line 823, in pandas._libs.tslib._array_to_datetime_object
    oresult[i] = parse_datetime_string(val, dayfirst=dayfirst,
  File "pandas/_libs/tslibs/parsing.pyx", line 264, in pandas._libs.tslibs.parsing.parse_datetime_string
    raise ValueError(f'Given date string {date_string} not likely a datetime.')
ValueError: Given date string 12 not likely a datetime.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/c/Users/danny/scratchpad/pandas/pandas/core/tools/datetimes.py", line 1064, in to_datetime
    values = convert_listlike(arg._values, format)
  File "/mnt/c/Users/danny/scratchpad/pandas/pandas/core/tools/datetimes.py", line 434, in _convert_listlike_datetimes
    result, tz_parsed = objects_to_datetime64ns(
  File "/mnt/c/Users/danny/scratchpad/pandas/pandas/core/arrays/datetimes.py", line 2206, in objects_to_datetime64ns
    result, tz_parsed = tslib.array_to_datetime(
  File "pandas/_libs/tslib.pyx", line 425, in pandas._libs.tslib.array_to_datetime
    cpdef array_to_datetime(
  File "pandas/_libs/tslib.pyx", line 680, in pandas._libs.tslib.array_to_datetime
    return _array_to_datetime_object(values, errors, dayfirst, yearfirst)
  File "pandas/_libs/tslib.pyx", line 832, in pandas._libs.tslib._array_to_datetime_object
    raise type(err)(f"Unable to parse string \"{val}\" at position {i}")
ValueError: Unable to parse string "12" at position 1
```

Reply:
i guess in one sentence, i mean to ask -
do we really need to invoke `_array_to_datetime_object` when the incorrect data exception has been caught earlier ? or could we consider avoiding it ?

Reply:
Probably, generally all object based arrays go though  `_array_to_datetime_object` and propagates through there. The date parsing logic is (unfortunately) complex but is also hard to change since branching somewhere else may have unintended side effects